<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Привет! Меня зовут Аня Титова, буду проверять твой проект по предобработке данных. 

Ты проделала достойную работу над проектом. Теперь осталось доработать несколько моментов, далее прокомментирую каждый из них.

Мои комментарии отмечены таким же заголовком «Комментарий ревьюера». Пожалуйста, не удаляй их и не перемещай – это важно для повторных проверок.
Красным цветом выделены замечания, которые нужно обязательно учесть.
Желтым выделены мои рекомендации, постарайся проработать их в этом проекте или в будущих.
Зеленым — элегантные решения, рекомендации «со звездочкой» и прочая информация.
</div>

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Привет. Спасибо за внимательную проработку замечаний! Дальше мои комментарии отмечены как **v2**.
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, обозначили цели исследования.
</div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [114]:
import pandas as pd
data=pd.read_csv('/datasets/data.csv')
data.info() #посмотрим общую информацию о данных

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-warning">
<b> Комментарий от ревьюера v2</b>

Заметила, что ячейки в твоей тетрадке теперь начинаются не с 1. Перед отправкой работы рекомендую перезапускать ноутбук, чтобы убедиться, что все ячейки выполняются корректно. Сделать это можно с помощью пункта меню Kernel > Restart & Run All

</div>

Все столбцы таблицы названы корректно, переименование не требуется. 
Есть пропуски в данных о трудовом стаже в днях и общем доходе, их доля существенна (10%) и количество совпадает. 
Удалить их и не принимать их в расчет при анализе данных нельзя.
Требуется работа с пропусками в данных и поиск дубликатов. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, заметили явные пропуски.
</div>

Просмотрим первые десять строк для наглядности.

In [115]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Просмотрим подробнее данные по столбцам, уникальные значения, наличие пропусков и т.д. 

In [116]:
print ('Основное:')
debt_values = data['debt'].unique()
print ('Данные о долгах:', *debt_values, sep='|')
children_values=data['children'].unique()
print ('Данные о детях:', *children_values, sep='|') 
family_status_values = data['family_status'].unique()
print ('Данные о семейном статусе:', *family_status_values, sep='|') 
print()
print('Проверка категорий')
gender_values = data['gender'].unique() 
print ('Данные о поле:', *gender_values, sep='|') 
income_type_values=data['income_type'].unique()
print ('Данные о видах занятости:', *income_type_values, sep='|') 
education_values = data['education'].unique()
print ('Данные об образовании:', *education_values, sep='|') 
print()
print('Проверка числовых столбцов')
print('Минимальное значение трудового стажа, лет: ', (data['days_employed'].min())/365, ', максимальное значение трудового стажа, лет: ', (data['days_employed'].max())/365, sep='')
print('Пропущенных значений о трудовом стаже: ', data['days_employed'].isna().sum())
print('Минимальное значение возраста, лет: ', data['dob_years'].min(), ', максимальное значение возраста, лет: ', data['dob_years'].max(), sep='')
print('Минимальное значение дохода: ', data['total_income'].min(), ', максимальное значение дохода: ', data['total_income'].max(), sep='')
print('Пропущенных значений о доходе: ', data['total_income'].isna().sum())

Основное:
Данные о долгах:|0|1
Данные о детях:|1|0|3|2|-1|4|20|5
Данные о семейном статусе:|женат / замужем|гражданский брак|вдовец / вдова|в разводе|Не женат / не замужем

Проверка категорий
Данные о поле:|F|M|XNA
Данные о видах занятости:|сотрудник|пенсионер|компаньон|госслужащий|безработный|предприниматель|студент|в декрете
Данные об образовании:|высшее|среднее|Среднее|СРЕДНЕЕ|ВЫСШЕЕ|неоконченное высшее|начальное|Высшее|НЕОКОНЧЕННОЕ ВЫСШЕЕ|Неоконченное высшее|НАЧАЛЬНОЕ|Начальное|Ученая степень|УЧЕНАЯ СТЕПЕНЬ|ученая степень

Проверка числовых столбцов
Минимальное значение трудового стажа, лет: -50.38068465909146, максимальное значение трудового стажа, лет: 1100.6997273296713
Пропущенных значений о трудовом стаже:  2174
Минимальное значение возраста, лет: 0, максимальное значение возраста, лет: 75
Минимальное значение дохода: 20667.26379327158, максимальное значение дохода: 2265604.028722744
Пропущенных значений о доходе:  2174


<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Отлично, наглядная проверка корректности значений. В будующих проектах можешь обернуть это код в функцию для категориальных и для количественных переменных, чтобы просто вызывать ее для интересующего столбца.</div>

In [117]:
#так может выглядеть функция для категориальных переменных
def category_function (column):
        list_name=column.unique()
        print ('Уникальные значения столбца: ', *list_name, sep='|')
        
#так может выглядеть функция для количественных переменных
def quantity_function (column): 
    print('Минимальное значение:', column.min(), ', максимальное значение:', column.max(), sep='')
    print('Пропущенных значений: ', column.isna().sum())

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Здорово, что сразу пробуешь составить такие функции. Для количественных переменных можно еще использовать метод describe().
</div>

###### Вывод

1 - Данные о семейном статусе заполнены все, пропусков нет, дубликатов в перечне уникальных значений нет. 

2 - Данные о видах занятости заполнены все, пропусков нет, дубликатов в перечне уникальных значений нет.  

3 - Данные о поле заполнены некорректно (есть третья величина).

4 - Данные о количестве детей содержат отрицательную величину.

5 - Данные об образовании содержат дубликаты в перечне уникальных значений. 

6 - Данные о трудовом стаже содержат некорректные данные (отрицательные).

7 - Данные о возрасте содержат как минимум нулевые данные, требуется дополнительное исследование, нет ли данных о возрасте менее 18 лет.
________________________________________________________________
Как определим факт погашения кредита в срок? Будем рассматривать эту задачу как нахождение доли клиентов с наличием / отсутствием долга (debt) среди общего числа клиентов по группам данных о семейном положении и количестве детей и искать взаимосвязь, сделав сводные таблицы, которые будут визуально пригодны для выводов. Мы никак не оцениваем взаимосвязь размера долга и длительности просрочек, т.к. размер долга в данных не указан, указан только факт его наличия (1) или отсутствия (0).

# Вывод

Требуется предобработка данных: обработка пропусков, дублей, выделение столбцов с категориями для формирования в дальнейшем сводных таблиц.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Отлично, есть первый взгляд на данные и намечены шаги для предобработки.</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

#### Исправим данные о поле клиента

Найдем количество строк с некорректным полом.

In [118]:
data.loc[data['gender']=='XNA']['gender'].count()

1

Т.к. это всего 1 строка из 21595, менее 0,005%, заменим данные пола этого клиента на мужской

In [119]:
data.loc[data['gender'] == 'XNA', 'gender'] = 'M'
data['gender'].unique()

array(['F', 'M'], dtype=object)

Некорректно заполненные данные о поле исправлены.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, есть обоснование замены и последующая проверка.
</div>

<div class="alert alert-warning">
Комментарии к действиям лучше давать в ячейках markdown — так их легче воспринимать. Комментарии в самом коде лучше использовать для пояснения технических аспектов, и только если они необходимы.
</div>


<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Вижу, что комментарии переехали в тектовые ячейки. Читать стало удобнее :) В целом лучше придерживаться такого принципа для всего проекта.</div>

#### Исправим данные о количестве детей

In [120]:
#найдем количество строк с количеством детей, равным "-1"
print(data.loc[data.loc[:, 'children']==-1]['children'].count())

47


In [121]:
#т.к. 47 из 21525 срок составляет лишь 0.2%, заменим данные с "-1" на "1", т.к. вероятнее всего, это техническая опечатка: 
data.loc[data['children'] == -1, 'children'] = 1
data['children'].unique() #проверим, как получилось

array([ 1,  0,  3,  2,  4, 20,  5])

Некорректно заполненные данные о количестве детей исправлены.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ок, замену отрицательных значений произвели и дали предположение, откуда они взялись. Для замены можно еще использовать функцию abs() https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.abs.html</div>

<div class="alert alert-warning">
Непонятно, что с 20 детьми - сколько таких записей, стоит ли их оставлять?
</div>

In [122]:
data.groupby('children')['children'].count()

children
0     14149
1      4865
2      2055
3       330
4        41
5         9
20       76
Name: children, dtype: int64

У клиента вполне может быть 20 детей, если это семейный детский дом, например. Таких записей 76 (0.35%), проверили ячейкой выше, их можно было бы и оставить, особенно с учетом того, что мы относим 20 детей к категории "более 3 детей" в дальнейшем. 
Однако признаем 20 детей опечаткой (20 вместо 2) и исправим код соответственно. 

In [123]:
data.loc[data['children']==20, 'children'] = 2
data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Хорошо, теперь есть два предположения о записях с 20 детьми и выбрано одно из них.
</div>

#### Проверим и исправим данные о возрасте клиентов

Мы знаем, что есть нулевые данные о возрасте, поищем иные ошибки - например, возраст менее 18 лет.

In [124]:
print ('Перечень возрастов клиентов:', data['dob_years'].unique())
print ('Количество клиентов с возрастом менее 18 лет: ', data.loc[data['dob_years']<18]['dob_years'].count())
print ('Количество клиентов с возрастом 0: ', data.loc[data['dob_years']==0]['dob_years'].count())

Перечень возрастов клиентов: [42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]
Количество клиентов с возрастом менее 18 лет:  101
Количество клиентов с возрастом 0:  101


Проверим медиану и среднее, чтобы понять, как лучше заполнить недостающие данные о возрасте.

In [125]:
dob_years_med=data['dob_years'].median()
dob_years_avg=data['dob_years'].mean()
print('Медиана: {:.0f}'.format(dob_years_med))
print('Среднее: {:.0f}'.format(dob_years_avg))

Медиана: 42
Среднее: 43


Т.к. 101 запись из 21525 это менее 0.5%, заменим эти данные на медиану, которая в данном случае почти равна среднему.

In [126]:
data.loc[data['dob_years']==0, 'dob_years']=dob_years_med
print ('Количество клиентов с возрастом менее 18 лет: ', data.loc[data['dob_years']<18]['dob_years'].count())
print ('Количество клиентов с возрастом 0: ', data.loc[data['dob_years']==0]['dob_years'].count())

Количество клиентов с возрастом менее 18 лет:  0
Количество клиентов с возрастом 0:  0


Данные о возрасте исправлены.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ок, возможный вариант обработки. Но есть вероятность, что у нас появились 42-летние студенты. В будующем постарайся обращать внимание на такие моменты.</div>

#### Проверим и исправим данные о трудовом стаже 

Посмотрев ранее на максимум и минимум по этому показателю, видим, что они некорректны: есть стаж, превышающий возраст в разы, есть отрицательный стаж. 

Проверим строки, стаж в которых превышает возраст. 

In [127]:
data.loc[data['days_employed']>(data['dob_years']*365)]['days_employed'].count

<bound method Series.count of 4        340266.072047
18       400281.136913
24       338551.952911
25       363548.489348
30       335581.668515
             ...      
21505    338904.866406
21508    386497.714078
21509    362161.054124
21518    373995.710838
21521    343937.404131
Name: days_employed, Length: 3445, dtype: float64>

Проверим, сколько строк, стаж в которых отрицателен.

In [128]:
data.loc[data['days_employed']<0]['days_employed'].count

<bound method Series.count of 0       -8437.673028
1       -4024.803754
2       -5623.422610
3       -4124.747207
5        -926.185831
            ...     
21519   -2351.431934
21520   -4529.316663
21522   -2113.346888
21523   -3112.481705
21524   -1984.507589
Name: days_employed, Length: 15906, dtype: float64>

В условиях того, что 3445+15906=19351 запись (а именно, все имеющиеся записи кроме NaN) либо превышают даже возраст клиента, либо отрицательны, а также учитывая то, что ответ на вопросы исследования не базируется на данных о трудовом стаже, заменим стаж на максимально возможный, а именно на величину возраста за вычетом 18 лет. 

In [129]:
data.loc[data['days_employed']>(data['dob_years']-18)*365, 'days_employed']=(data['dob_years']-18)*365
data.loc[data['days_employed']<0, 'days_employed']=(data['dob_years']-18)*365
data['days_employed']=data['days_employed'].fillna(0)

In [130]:
print('Количество пропусков в данных о трудовом стаже: ', data['days_employed'].isna().sum())

Количество пропусков в данных о трудовом стаже:  0


Скорректировали данные о трудовом стаже, хотя это и те данные, которые мы не используем при ответе на вопросы исследования.

<div class="alert alert-warning">
<b> Комментарий ревьюера</b>

Ок, возможный вариант. Но так как мы здесь заменили все записи датафрейма, лучше сделать предположение, откуда эти аномалии, и почему стоит обработать их именно так.
</div>

#### Проверим и исправим данные о доходе

Этот параметр мы используем при ответе на вопросы исследования. 

In [44]:
print('Пропущенных значений о доходе: ', data['total_income'].isna().sum())

Пропущенных значений о доходе:  2174


Найдем медиану дохода по виду деятельности (income_type).

In [131]:
data_grouped_by_income_type=data.groupby('income_type')['total_income'].median()
print(data_grouped_by_income_type)

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64


Сделаем из перечня уникальных значений видов занятости список.

In [132]:
import numpy as np
income_types_list=np.array(data['income_type'].unique()).tolist()
income_types_list 

['сотрудник',
 'пенсионер',
 'компаньон',
 'госслужащий',
 'безработный',
 'предприниматель',
 'студент',
 'в декрете']

Пройдемся по списку и для каждого вида деятельности заменим пропущенное значение в столбце total_income на медиану по группе.

In [133]:
for name in income_types_list: 
    data.loc[data['income_type']==name,'total_income']=data.loc[data['income_type']==name,'total_income'].fillna(data.loc[data['income_type']==name,'total_income'].median())

In [134]:
data['total_income'].isna().sum()

0

Проверим теперь медиану дохода по виду деятельности (income_type).

In [136]:
data_grouped_by_income_type=data.groupby('income_type')['total_income'].median()
print(data_grouped_by_income_type)

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64


Данные о доходах исправили на основании медианных значений по виду деятельности клиента. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Отлично, что учитываешь род деятельности для заполнения пропусков. Элегантное решение в коде.

На что еще можно обращать внимание: свзяаны ли были эти пропуски с другими пропусками в данных.
</div>

По количеству пропуски о доходе равны пропуска в трудовом стаже. Вряд ли возможно, что при получении кредита клиент не указал трудовой стаж и не указал свой доход, и все равно был одобрен как заемщик. Возможно, эти данные должны были быть подгружены из справок, но подгружены не были?

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Ок, возможный вариант. В реальной работе с такими предположениями можно обращаться к технической команде.
</div>

### Замена типа данных

In [138]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null float64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 2.0+ MB


Превратим количество дней стажа и возраст клиента в целое число.

In [139]:
data['days_employed']=data['days_employed'].astype('int')
data['dob_years']=data['dob_years'].astype('int')

In [140]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


**Вывод**

Типы данных приведены в соответствие с логикой, числа с плавающей запятой в кол-ве дней стажа и возраста переведены в целые. Строки (категории) имеют строковый тип данных, количественные значения - int, float. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, привели типы. 
    
Комментарий о том, что по логике некоторые столбцы должны быть целочисленными, лучше помещать после шага, когда увидели эту проблему, и до шага, когда проблему исправили.

</div>

### Обработка дубликатов

Есть дубликаты в названиях видов образования.

In [142]:
print(data['education'].unique())

['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень']


In [145]:
#можно все их вручную переписать
#data.loc[data['education']=='Среднее', 'education']='среднее'
#data.loc[data['education']=='СРЕДНЕЕ', 'education']='среднее'
#data.loc[data['education']=='ВЫСШЕЕ', 'education']='высшее'
#data.loc[data['education']=='Высшее', 'education']='высшее'
#data.loc[data['education']=='НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'education']='неоконченное высшее'
#data.loc[data['education']=='Неоконченное высшее', 'education']='неоконченное высшее'
#data.loc[data['education']=='НАЧАЛЬНОЕ', 'education']='начальное'
#data.loc[data['education']=='Начальное', 'education']='начальное'
#data.loc[data['education']=='Ученая степень', 'education']='ученая степень'
#data.loc[data['education']=='УЧЕНАЯ СТЕПЕНЬ', 'education']='ученая степень'

In [146]:
#print(data['education'].unique())

<div class="alert alert-warning">
<b> Комментарий ревьюера</b>
        
Да, от дубликатов в образовании избавились. Но стоит сделать предположение о том, откуда они могли взяться. Еще в данных есть id для каждого вида образования — постарайся учесть и эту информацию.
    
Для оптимизации кода здесь можно использовать функцию str.lower() — она работает сразу со всем столбцом датафрейма https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.lower.html
</div>

Дубликаты в названии образования взялись, вероятно, оттого, что в анкете это поле заполняется вручную, а не выбирается одна из категорий выпадающего списка, поэтому каждый клиент вводил данные так, как хотел. Оптимизируем код с учетом замечания, закомментировав в ячейках выше исправления, внесенные вручную. Проверим результат. 

In [148]:
data['education']=data['education'].str.lower()
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Устранили дубликаты в названиях видов образования. Также это можно было сделать через замену значения ячейки, используя как базу education_id.

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Хорошо, принято.
</div>

Проверим наличие полных дубликатов строк.

In [149]:
print(data.duplicated().sum())

72


Доля дубликатов составляет 0,33% от всех данных, и вероятность полного совпадения настоящих клиентов друг с другом по такому большому числу параметров очень мала. 
Дубликаты можно удалить. Причина появления дубликатов, возможно, повторное занесение анкет клиентов или случайная повторная отправка (сохранение) их анкет. 

In [150]:
data=data.drop_duplicates().reset_index(drop=True)
print(data.duplicated().sum())

0


In [151]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 12 columns):
children            21453 non-null int64
days_employed       21453 non-null int64
dob_years           21453 non-null int64
education           21453 non-null object
education_id        21453 non-null int64
family_status       21453 non-null object
family_status_id    21453 non-null int64
gender              21453 non-null object
income_type         21453 non-null object
debt                21453 non-null int64
total_income        21453 non-null float64
purpose             21453 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


**Вывод**

В данных нет пропусков и дубликатов. Типы данных приведены в соответствие с их содержанием. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо.

</div>

### Лемматизация

Просмотрим, сколько вариантов целей взятия кредита содержится в данных.

In [152]:
purpose_list=data['purpose'].unique().tolist()
purpose_list

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

Узнаем их количество.

In [153]:
len(purpose_list)

38

В 38 разных вариантах видим 4 основных категории: недвижимость, автомобиль, свадьба, образование

In [154]:
def purpose_function (p): 
    if 'жил' in p:
        return 'недвижимость'
    if 'недвиж' in p:
        return 'недвижимость'
    if 'автом' in p: 
        return 'автомобиль'
    if 'свад' in p: 
        return 'свадьба'
    if 'образ' in p: 
        return 'образование'
data['purpose_category']=data['purpose'].apply(purpose_function)
print(data['purpose_category'].value_counts())

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2323
Name: purpose_category, dtype: int64


<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо.Понятно, как выделяли категории.
</div>

Попробуем то же самое с леммами. 

In [156]:
from pymystem3 import Mystem
m = Mystem()
def purpose_function(row): 
    lemmas_purpose=m.lemmatize(row['purpose'])
    return lemmas_purpose

data['lemmas_of_purpose']=data.apply(purpose_function, axis=1)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose  
0                    [покупка,  , жилье, \n]  
1          [приобретение,  , автомобиль, \n]  
2                    [покупка,  , жилье, \n]  
3       [дополнительный,  , образование, \n]  
4                  [сыграть,  , свадьба, \n]  
5                    [покупка,  , жилье, \n]  
6             [операция,  , с,  , жилье, \n]  
7                          [образование, \n]  
8        [на,  , проведение,  , свадьба, \n]  
9  [покупка,  , жилье,  , для,  , семья, \n]

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, провели лемматизацию.

</div>

Леммы получились. На основании их также можно отнести цель кредита к одной из четырех категорий, к которым мы уже отнесли их выше (получившийся столбец purpose_category). 

In [157]:
def purpose_function_lemmas (p): 
    if 'жилье' in p:
        return 'недвижимость'
    if 'недвижимость' in p:
        return 'недвижимость'
    if 'автомобиль' in p: 
        return 'автомобиль'
    if 'свадьба' in p: 
        return 'свадьба'
    if 'образование' in p: 
        return 'образование'
data['purpose_category_lemmas']=data['lemmas_of_purpose'].apply(purpose_function_lemmas)
print(data['purpose_category_lemmas'].value_counts())

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2323
Name: purpose_category_lemmas, dtype: int64


In [69]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose purpose_category_lemmas  
0                    [покупка,  , жилье, \n]            недвижимость  
1          [приобретение,  , автомобиль, \n]              автомобиль  
2                    [покупка,  , жилье, \n]            недвижимость  
3       [дополнительный,  , образование, \n]             образование  
4                  [сыграть,  , свадьба, \n]                 свадьба  
5                    [покупка,  , жилье, \n]            недвижимость  
6             [операция,  , с,  , жилье, \n]            недвижимость  
7                          [образование, \n]             образование  
8        [на,  , проведение,  , свадьба, \n]                 свадьба  
9  [покупка,  , жилье,  , для,  , семья, \n]            недвижимость

**Вывод**

Второй способ показал то же самое. Категории, выделенные на основе лемм и ручным способом выше, совпали. Данные получили категории согласно целям получения кредита. Их мы используем для сводных таблиц и ответа на вопрос, зависит ли просрочки по кредитам от целей взятия кредита. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Здорово, что сравниваешь два способа категоризации.
    
В целом на этом шаге было достаточно выделить леммы, а разбить по категориям можно на следующем шаге. Если бы категорий было больше, или они были бы не такие явные, можно выделять еще категорию «остальное».

</div>

### Категоризация данных

In [158]:
print(data['children'].value_counts())

0    14090
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64


Разброс довольно велик, поэтому отнесем клиента к одной из категорий, которую мы введем на основании данных о том, сколько у него детей, это пригодится при создании сводных таблиц

In [159]:
def group_function (qty): 
    if qty==0:
        return 'нет детей'
    elif qty==1:
        return '1 ребенок'
    elif 2<=qty<=3: 
        return '2-3 ребенка'
    elif qty>3:
        return 'более 3 детей'
data['children_category']=data['children'].apply(group_function)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose purpose_category_lemmas  \
0                    [покупка,  , жилье, \n]            недвижимость   
1          [приобретение,  , автомобиль, \n]              автомобиль   
2                    [покупка,  , жилье, \n]            недвижимость   
3       [дополнительный,  , образование, \n]             образование   
4                  [сыграть,  , свадьба, \n]                 свадьба   
5                    [покупка,  , жилье, \n]            недвижимость   
6             [операция,  , с,  , жилье, \n]            недвижимость   
7                          [образование, \n]             образование   
8        [на,  , проведение,  , свадьба, \n]                 свадьба   
9  [покупка,  , жилье,  , для,  , семья, \n]            недвижимость   

  children_category  
0         1 ребенок  
1         1 ребенок  
2         нет детей  
3       2-3 ребенка  
4         нет детей  
5         нет детей  
6         нет детей  
7         нет детей  
8       2-3 ребенка  
9         нет детей

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, распределили по категориям. 
</div>    
<div class="alert alert-warning">
Возможно, здесь что-то изменится после предобработки записей с 20 детьми.

</div>

Данные о 20 детях были изменены, поэтому и вывод изменился.

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Ок.</div>

Посмотрим, сколько у нас клиентов разного возраста.

In [160]:
print(data['dob_years'].value_counts())

42    696
35    616
40    607
41    605
34    601
38    597
33    581
39    572
31    559
36    554
44    545
29    544
30    537
37    536
48    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
55    443
59    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


Разброс велик, поэтому отнесем клиента к одной из категорий, которую мы введем на основании данных о его возрасте, это пригодится при создании сводных таблиц. 

In [161]:
def age_selection (age): 
    if 18<=age<=30:
        return '18-30 лет'
    elif 30<age<=45:
        return '31-45 лет'
    elif 45<age<=60:
        return '46-60 лет'
    elif age>60:
        return 'более 60 лет'
data['age_group']=data['dob_years'].apply(age_selection)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose purpose_category_lemmas  \
0                    [покупка,  , жилье, \n]            недвижимость   
1          [приобретение,  , автомобиль, \n]              автомобиль   
2                    [покупка,  , жилье, \n]            недвижимость   
3       [дополнительный,  , образование, \n]             образование   
4                  [сыграть,  , свадьба, \n]                 свадьба   
5                    [покупка,  , жилье, \n]            недвижимость   
6             [операция,  , с,  , жилье, \n]            недвижимость   
7                          [образование, \n]             образование   
8        [на,  , проведение,  , свадьба, \n]                 свадьба   
9  [покупка,  , жилье,  , для,  , семья, \n]            недвижимость   

  children_category  age_group  
0         1 ребенок  31-45 лет  
1         1 ребенок  31-45 лет  
2         нет детей  31-45 лет  
3       2-3 ребенка  31-45 лет  
4         нет детей  46-60 лет  
5         нет детей  18-30 лет  
6         нет детей  31-45 лет  
7         нет детей  46-60 лет  
8       2-3 ребенка  31-45 лет  
9         нет детей  31-45 лет

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ок, возможный вариант разбивки. Можно еще указывать, как именно выбирали границы возраста для категорий.

</div>

Проверим семейное положение клиентов. Сделаем более крупную категорию. Отнесем клиента к одной из двух категорий, которую мы введем, на основании данных о том, одинок он или находится в паре. 

In [162]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [164]:
def group_family_status (status): 
    if status=='женат / замужем' or status=='гражданский брак':
        return 'в паре'
    return 'одинок'
data['family_status_category']=data['family_status'].apply(group_family_status) 
data.head(10) 

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose purpose_category_lemmas  \
0                    [покупка,  , жилье, \n]            недвижимость   
1          [приобретение,  , автомобиль, \n]              автомобиль   
2                    [покупка,  , жилье, \n]            недвижимость   
3       [дополнительный,  , образование, \n]             образование   
4                  [сыграть,  , свадьба, \n]                 свадьба   
5                    [покупка,  , жилье, \n]            недвижимость   
6             [операция,  , с,  , жилье, \n]            недвижимость   
7                          [образование, \n]             образование   
8        [на,  , проведение,  , свадьба, \n]                 свадьба   
9  [покупка,  , жилье,  , для,  , семья, \n]            недвижимость   

  children_category  age_group family_status_category  
0         1 ребенок  31-45 лет                 в паре  
1         1 ребенок  31-45 лет                 в паре  
2         нет детей  31-45 лет                 в паре  
3       2-3 ребенка  31-45 лет                 в паре  
4         нет детей  46-60 лет                 в паре  
5         нет детей  18-30 лет                 в паре  
6         нет детей  31-45 лет                 в паре  
7         нет детей  46-60 лет                 в паре  
8       2-3 ребенка  31-45 лет                 в паре  
9         нет детей  31-45 лет                 в паре

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, иногда стоит объединять похожие или малочисленные группы. Красивое решение с функцией.

</div>

Отнесем клиента к одной из категорий в зависимости от размера дохода, который он получает. 

In [165]:
quantile25percent = data['total_income'].quantile([0.25], interpolation='nearest')
median_data=data['total_income'].median()
quantile75percent = data['total_income'].quantile([0.75], interpolation='nearest')
print ('Первая четверть:', quantile25percent)
print ('Медиана:', median_data)
print ('Вторая четверть:', quantile75percent)

Первая четверть: 0.25    107620.948105
Name: total_income, dtype: float64
Медиана: 142594.39684740017
Вторая четверть: 0.75    195821.632457
Name: total_income, dtype: float64


In [166]:
def income_rate (income): 
    if income<=int(quantile25percent):
        return 'низкий доход'
    if int(quantile25percent)<income<=int(median_data):
        return 'средний доход'
    if int(median_data)<income<=int(quantile75percent):
        return 'высокий доход'
    if income>int(quantile75percent):
        return 'очень высокий доход'
data['income_group']=data['total_income'].apply(income_rate)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8760         42    высшее             0   
1         1           6570         36   среднее             1   
2         0           5475         33   среднее             1   
3         3           5110         32   среднее             1   
4         0          12775         53   среднее             1   
5         0           3285         27    высшее             0   
6         0           9125         43    высшее             0   
7         0          11680         50   среднее             1   
8         2           6205         35    высшее             0   
9         0           8395         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  \
0               покупка жилья     недвижимость   
1     приобретение автомобиля       автомобиль   
2               покупка жилья     недвижимость   
3  дополнительное образование      образование   
4             сыграть свадьбу          свадьба   
5               покупка жилья     недвижимость   
6           операции с жильем     недвижимость   
7                 образование      образование   
8       на проведение свадьбы          свадьба   
9     покупка жилья для семьи     недвижимость   

                           lemmas_of_purpose purpose_category_lemmas  \
0                    [покупка,  , жилье, \n]            недвижимость   
1          [приобретение,  , автомобиль, \n]              автомобиль   
2                    [покупка,  , жилье, \n]            недвижимость   
3       [дополнительный,  , образование, \n]             образование   
4                  [сыграть,  , свадьба, \n]                 свадьба   
5                    [покупка,  , жилье, \n]            недвижимость   
6             [операция,  , с,  , жилье, \n]            недвижимость   
7                          [образование, \n]             образование   
8        [на,  , проведение,  , свадьба, \n]                 свадьба   
9  [покупка,  , жилье,  , для,  , семья, \n]            недвижимость   

  children_category  age_group family_status_category         income_group  
0         1 ребенок  31-45 лет                 в паре  очень высокий доход  
1         1 ребенок  31-45 лет                 в паре        средний доход  
2         нет детей  31-45 лет                 в паре        высокий доход  
3       2-3 ребенка  31-45 лет                 в паре  очень высокий доход  
4         нет детей  46-60 лет                 в паре        высокий доход  
5         нет детей  18-30 лет                 в паре  очень высокий доход  
6         нет детей  31-45 лет                 в паре  очень высокий доход  
7         нет детей  46-60 лет                 в паре        средний доход  
8       2-3 ребенка  31-45 лет                 в паре         низкий доход  
9         нет детей  31-45 лет                 в паре        высокий доход

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Отлично. Поделив на группы по процентилям, мы сделали их равными по объему.

</div>

Мы создали категории на основании данных о количестве детей, возрасте, семейном положении и доходе. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Да, категоризацию выполнили.

</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Посмотрим, как количество детей влияет на возврат кредита, используя данные о полном количестве детей без категорий. Данные изменились с учетом того, что мы исправили 20 детей на 2, признав это за опечатку. 

Построим сводную таблицу о наличии / отсутствии долгов и количестве детей.

In [167]:
children_pivot_table=data.pivot_table(index=['children'], columns='debt', values='dob_years', aggfunc='count')
children_pivot_table

debt            0       1
children                 
0         13027.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

Обнаружили NaN как данные о тех, у кого пять детей и есть долги, т.е. тех клиентов, у кого 5 детей и есть долг, нет вообще. Заменим на ноль. Отсортируем по столбцу уровня задержки кредита эту и все последующие таблицы. 

In [168]:
children_pivot_table=children_pivot_table.fillna(0)
children_pivot_table['debt_ratio_percent']=(children_pivot_table[1]/children_pivot_table[0])*100
children_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt            0       1  debt_ratio_percent
children                                     
5             9.0     0.0            0.000000
0         13027.0  1063.0            8.159975
3           303.0    27.0            8.910891
1          4410.0   445.0           10.090703
2          1926.0   202.0           10.488058
4            37.0     4.0           10.810811

Учитывая количество данных в той или иной группе, видим, что зависимость есть: чем больше детей, тем выше уровень задержки кредита, хотя на первом (самом благонадежном месте) клиенты с 5 детьми - но их количество очень мало: всего 9 человек. 
Для наглядности построим еще одну сводную таблицу, используем категории, введенные нами ранее в отношении количества детей. 

In [169]:
children_category_pivot_table=data.pivot_table(index=['children_category'], columns='debt', values='dob_years', aggfunc='count')
children_category_pivot_table['debt_ratio_percent']=(children_category_pivot_table[1]/children_category_pivot_table[0])*100
children_category_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                   0     1  debt_ratio_percent
children_category                                 
нет детей          13027  1063            8.159975
более 3 детей         46     4            8.695652
1 ребенок           4410   445           10.090703
2-3 ребенка         2229   229           10.273665

**Вывод**

Видим, что чем больше детей, тем выше показатель задержки или невозврата кредита, показатель debt_ratio_percent возрастает. Клиенты без детей в меньшей степени задерживают / не выплачивают кредиты. 
Зависимость есть, она проиллюстрирована в этой сводной таблице.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, посмотрели на связь с количеством детей с разным делением по категориям. Отличное применение сводных таблиц.
</div>

<div class="alert alert-warning">
    
В выводе стоит прямо указать, какую именно зависимость обнаружили.

</div>

<div class="alert alert-success">
<b> Комментарий ревьюера v2</b>

Отлично, внесли ясность в вывод.
</div>

Обнаруженная зависимость прямая: чем больше детей, тем больше уровень наличия просрочки по кредиту. 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [170]:
family_status_pivot_table=data.pivot_table(index=['family_status'], columns='debt', values='dob_years', aggfunc='count')
family_status_pivot_table['debt_ratio_percent']=(family_status_pivot_table[1]/family_status_pivot_table[0])*100
family_status_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                       0    1  debt_ratio_percent
family_status                                        
вдовец / вдова           896   63            7.031250
в разводе               1110   85            7.657658
женат / замужем        11408  931            8.160940
гражданский брак        3762  388           10.313663
Не женат / не замужем   2536  274           10.804416

Клиенты, не состоящие в паре (и не бывшие женаты /замужем) или состоящие в гражданском браке, в большей степени задерживают кредиты, чем те, кто женат/замужем, в разводе или вдовец. Проверим на основании более укрупненной категории. 

In [171]:
family_status_category_pivot_table=data.pivot_table(index=['family_status_category'], columns='debt', values='dob_years', aggfunc='count')
family_status_category_pivot_table['debt_ratio_percent']=(family_status_category_pivot_table[1]/family_status_category_pivot_table[0])*100
family_status_category_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                        0     1  debt_ratio_percent
family_status_category                                 
в паре                  15170  1319            8.694792
одинок                   4542   422            9.291061

Исследовав данные также на основании категории одинок клиент или состоит в паре, мы видим, что клиент, не состоящий в паре, с большей вероятностью может иметь просрочку или невозврат по кредиту. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошо, ответили на вопрос.

</div>

Также посмотрим, как возраст влияет на просрочки по кредитам. 

In [85]:
age_group_pivot_table=data.pivot_table(index=['age_group'], columns='debt', values='dob_years', aggfunc='count')
age_group_pivot_table['debt_ratio_percent']=(age_group_pivot_table[1]/age_group_pivot_table[0])*100
age_group_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt             0    1  debt_ratio_percent
age_group                                  
более 60 лет  2025  101            4.987654
46-60 лет     6557  467            7.122160
31-45 лет     7816  770            9.851586
18-30 лет     3314  403           12.160531

Чем выше возраст, тем меньшая доля наличия просрочек по кредиту.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ок.

</div>

Посмотрим по двум категориям одновременно: по возрасту и семейному положению.

In [172]:
two_group_pivot_table=data.pivot_table(index=['family_status_category','age_group'], columns='debt', values='dob_years', aggfunc='count')
two_group_pivot_table['debt_ratio_percent']=(two_group_pivot_table[1]/two_group_pivot_table[0])*100
two_group_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                                    0    1  debt_ratio_percent
family_status_category age_group                                  
в паре                 более 60 лет  1387   64            4.614275
одинок                 более 60 лет   638   37            5.799373
в паре                 46-60 лет     5051  353            6.988715
одинок                 46-60 лет     1506  114            7.569721
в паре                 31-45 лет     6424  631            9.822540
одинок                 31-45 лет     1392  139            9.985632
в паре                 18-30 лет     2308  271           11.741768
одинок                 18-30 лет     1006  132           13.121272

Здесь мы также видим, что чем выше возраст, тем меньше просрочек по кредиту, и в каждой категории также видим, что люди, состоящие в паре в меньшей степени имеют просрочки, чем те, кто одинок.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Хорошее замечание.

</div>

Сделаем сводную таблицу, которая показывала бы зависимость просрочки по кредиту от семейного положения и количества детей одновременно.

In [173]:
two_group_pivot_table=data.pivot_table(index=['children_category', 'family_status_category'], columns='debt', values='dob_years', aggfunc='count')
two_group_pivot_table['debt_ratio_percent']=(two_group_pivot_table[1]/two_group_pivot_table[0])*100
two_group_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                                         0    1  debt_ratio_percent
children_category family_status_category                               
более 3 детей     в паре                    43    3            6.976744
нет детей         в паре                  9452  745            7.881930
                  одинок                  3575  318            8.895105
1 ребенок         в паре                  3639  365           10.030228
2-3 ребенка       в паре                  2036  206           10.117878
1 ребенок         одинок                   771   80           10.376135
2-3 ребенка       одинок                   193   23           11.917098
более 3 детей     одинок                     3    1           33.333333

Одинокие люди с числом детей более 3 в два раза больше допускают просрочек по выплате кредитов, чем люди, находящиеся в паре (в браке, в гражданском браке) с тем же количеством детей. Меньше всего просрочек допускают клиенты, состоящие в паре и не имеющие детей, а также клиенты, не состоящие в паре и не имеющие детей.

###### Вывод

Чем больше детей, тем больше просрочек по кредиту. А наличие пары имеет положительное влияние на платежеспособность. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Здорово, что смотришь на данные под разными углами.

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [174]:
income_pivot_table=data.pivot_table(index=['income_group'], columns='debt', values='dob_years', aggfunc='count')
income_pivot_table['debt_ratio_percent']=(income_pivot_table[1]/income_pivot_table[0])*100
income_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                    0    1  debt_ratio_percent
income_group                                      
очень высокий доход  4981  383            7.689219
низкий доход         4936  427            8.650729
высокий доход        5770  546            9.462738
средний доход        4025  385            9.565217

Клиенты с очень высоким доходом в меньшей степени допускают просрочки по кредиту, как и клиенты с низким доходом. Больше просрочек допускают те, кто имеет высокий и средний доход. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ок.

</div>

- Как разные цели кредита влияют на его возврат в срок?

In [175]:
purpose_pivot_table=data.pivot_table(index=['purpose_category'], columns='debt', values='dob_years', aggfunc='count')
purpose_pivot_table['debt_ratio_percent']=(purpose_pivot_table[1]/purpose_pivot_table[0])*100
purpose_pivot_table.sort_values(by='debt_ratio_percent', ascending=True)

debt                  0    1  debt_ratio_percent
purpose_category                                
недвижимость      10029  782            7.797388
свадьба            2137  186            8.703790
образование        3643  370           10.156464
автомобиль         3903  403           10.325391

Просрочек за кредит, взятый на автомобиль и образование, больше, чем на свадьбу и недвижимость. 
Просрочек по кредитам на недвижимость меньше всего.

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Ответили на вопрос.

</div>

## Шаг 4. Общий вывод

Мы нашли зависимость просрочек по кредитам от количества детей, уровня дохода, целей кредита и семейного положения клиентов. 
Меньше всего просрочек по кредитам на недвижимость, по кредитам, взятым людьми более старшего возраста, людьми, состоящими в браке (гражданском браке) и не имеющими детей. 
Клиенты с самым высоким доходом и с самым низким доходом задерживают кредит в меньшей степени, чем клиенты со средним доходом. 

<div class="alert alert-success">
<b> Комментарий ревьюера</b>

Отлично, ответили на вопросы исследования.
    
Дополнительно в выводе можно указать, с какими данными работали, с какими проблемами столкнулись. Помимо рекомендаций по кредитному скорингу твои выводы могут содержать ценную информацию для технической команды, которая готовит выгрузки или собирает данные.

</div>

<div class="alert alert-warning">
<b> Итоговый комментарий ревьюера</b>

Отличная работа, молодец! Проект неплохо оформлен, сделаны основные шаги, есть промежуточные выводы и итоговые рекомендации. Критичных замечаний у меня нет, но мелких многовато. Постарайся, пожалуйста, учесть хотя бы несколько из них, чтобы я могла зачесть проект. Напомню, это было в желтых комментариях:

* Комментарии к действиям в ячейках markdown, не в коде
* Обработка записей с 20 детьми
* Гипотезы о возникновении аномалий в данных
* Гипотезы о возникновении дубликатов
* Использование готовых функций Pandas
* Мелочи в выводах на 3 шаге

Зеленые комментарии содержат инфомацию на будущее, в этом проекте учитывать их не обязательно.

Успехов!
</div>

Все комментарии перенесла в ячейки, убрав из кода ненужные пояснения. 
Обработка записей с 20 детьми: провела по-другому, хотя это можно было и обосновать (возможно, это семейный детский дом). 
Гипотезы о возникновении аномалий в данных: внесла (не подгружаются справки, анкеты). 
Гипотезы о возникновении дубликатов: внесла (дублирование анкет). 
Мелочи в выводах на 3 шаге: указала, что найденная зависимость прямая. 
СПАСИБО!

<div class="alert alert-success">
<b> Итоговый комментарий ревьюера v2</b>

Просмотрела изменения, отличная работа! Твой проект принят.
    
Еще два замечания на будущее:
* Постарайся перезапускать ядро и все ячейки перед отправкой проекта. Это придает тетрадке более опрятный вид и позволяет отловить ошибки кода.
* Когда меняешь или удаляешь значительную часть данных, лучше приводить подробные обоснования для такого действия.
    
Спасибо за классный проект, было приятно его проверять.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.